### PR TITLE
Accounting for zero-snap giving infinite likelihood in match.py

### PR DESCRIPTION
**Pull request by tomasdsousa**
Implementing in match.py the check already implemented in test_all_Fisher.py in issue #20, which checks whether it's fine to set parameters to zero when parameter/delta<1, or if doing that will return an infinite likelihood. To check it works we use the example likelihood class from that issue:

`class MyLikelihood(Likelihood):
    
    def __init__(self, nx):
        super().__init__('data_file', 'cov_file', 'core_maths',)
        self.ylabel = r'$y$'
        self.xvar = np.random.uniform(0, 1, size=nx)
        self.yerr = np.full(nx, 0.2)
        self.yvar = np.ones(nx) + self.yerr * np.random.normal(0, 1, size=nx)
        
    def negloglike(self, a, eq_numpy, **kwargs):
        try:
            if a[1] <= 0:
                return np.inf
        except:
            pass
        ypred = self.get_pred(self.xvar, np.atleast_1d(a), eq_numpy)
        if not np.all(np.isreal(ypred)):
            return np.inf
        nll = np.sum(0.5 * (ypred - self.yvar) ** 2 / self.yerr ** 2 + 0.5 * np.log(2 * np.pi) + np.log(self.yerr))
        if np.isnan(nll):
            return np.inf
        return nll`

Now, running ESR for complexity 5 using the core_maths function library, functions such as pow(Abs(a0),(pow(Abs(a1),x))) make it to the final list. Previosuly they would but with inf DL, even correcting what seemed to me to be a mistake in the match.py file that was leading to the nll to be set to nan for some functions, namely nparam in lines 129 and 141 which should be nparams.

### CODE REVIEW

File: esr/fitting/match.py [hunk 1]
```diff
@@ -147,17 +152,70 @@ def f1(x):
...
+                        if np.isfinite(negloglike_all[i]):
+                            break
+                kept_mask = np.ones(len(p), dtype=bool)
+                if np.isfinite(negloglike_all[i]):
+                    k -= len(idx)
+                    kept_mask[idx] = 0
+                elif not np.isfinite(negloglike_all[i]) and not np.isnan(negloglike_all[i]): # infinite nll
+                    p = ptrue
+                    fish[Nsteps<1] = 12./(p[Nsteps<1]**2) # set uncertainty=parameter in this case
+                    codelen[i] = -k/2.*math.log(3.) + np.sum( 0.5*np.log(fish_special) + np.log(abs(np.array(p))) )
```
Comment (DeaglanBartlett) [inline esr/fitting/match.py:184]: fish_special does not seem to be defined anywhere, please do this

**Review by DeaglanBartlett — CHANGES_REQUESTED** [review]

File: esr/fitting/match.py (hunk 1 above)
Comment (tomasdsousa) [reply esr/fitting/match.py]: Yes sorry it should be "fish", I forgot to upload that part from my laptop

File: esr/fitting/match.py [hunk 2]
```diff
@@ -147,17 +152,70 @@ def f1(x):
...
+                kept_mask = Nsteps>=1
+            else:
+                # Let's see if setting any of the parameters to zero is ok
+                try_idx = np.arange(nparams)[Nsteps < 1]
+                for r in reversed(range(1, len(try_idx))):
+                    for idx in itertools.combinations(try_idx, r):
+                        p = np.copy(ptrue)
+                        p[idx] = 0.
+                        if k==1:
+                            eq_numpy = sympy.lambdify([x, a0], eq, modules=["numpy"])
```
Comment (DeaglanBartlett) [inline esr/fitting/match.py:167]: This is unnecessary as we are still on the same function, so we don't need to re-sympify the equation as eq_numpy is already the one we're after. Please delete this line.

File: esr/fitting/match.py [hunk 3]
```diff
@@ -147,17 +152,70 @@ def f1(x):
...
+                            negloglike_all[i] = f1(p)               # Modified here for this variant, but if this doesn't happen it stays the same as the unique eq
+                        else:
+                            all_a = ' '.join([f'a{i}' for i in range(nparams)])
+                            all_a = list(sympy.symbols(all_a, real=True))
+                            eq_numpy = sympy.lambdify([x] + all_a, eq, modules=["numpy"])
```
Comment (DeaglanBartlett) [inline esr/fitting/match.py:172]: Same as previous comment - can be deleted

File: esr/fitting/match.py [hunk 4]
```diff
@@ -147,17 +152,70 @@ def f1(x):
...
+                            negloglike_all[i] = fop(p)
+                        # negloglike_all[i] = f1(p)
```
Comment (DeaglanBartlett) [inline esr/fitting/match.py:174]: Please deleted unused commented line

**Review by DeaglanBartlett — CHANGES_REQUESTED** [review]

File: esr/fitting/match.py [hunk 5]
```diff
@@ -178,9 +178,9 @@ def f1(x):
                 if np.isfinite(negloglike_all[i]):
                     k -= len(idx)
                     kept_mask[idx] = 0
-                else: # infinite/nan nll
+                elif not np.isfinite(negloglike_all[i]) and not np.isnan(negloglike_all[i]): # infinite nll
                     p = ptrue
-                    fish_special = 12./(p**2) # set uncertainty=parameter in this case
+                    fish[Nsteps<1] = 12./(p[Nsteps<1]**2) # set uncertainty=parameter in this case
                     codelen[i] = -k/2.*math.log(3.) + np.sum( 0.5*np.log(fish_special) + np.log(abs(np.array(p))) )
```
Comment (DeaglanBartlett) [inline esr/fitting/match.py]: Is this duplicated by the step at line 213? If so, we don't need to do it twice

**Review by DeaglanBartlett — APPROVED** [review]

**Pull request merged**